pkg/maintainer/spv/spv.go

```
@@ -59,8 +63,271 @@ func (sm *spvMaintainer) startControlLoop(ctx context.Context) {
+			// proven or it's not a deposit sweep transaction at all.
+			if !isMainUtxo {
+				return false, nil
+			}
+
+			// The input is the current main UTXO of the wallet. Proceed with
+			// checking other inputs.
+		} else {
+			// The input is a deposit input. Check if it swept or not.
+			if deposit.SweptAt.Equal(time.Unix(0, 0)) {
+				// The input is a deposit and it's unswept.
+				hasDepositInputs = true
+			} else {
+				// The input is a deposit, but it's already swept.
+				// The transaction must a deposit sweep transaction, but it's
+				// already proven.
+				return false, nil
+			}
+		}
+	}
+
+	// All the inputs represent either unswept deposits or the current main UTXO.
+	// As the final check verify if at least one of them was a deposit input.
+	// This will distinguish a deposit sweep transaction from a different
+	// transaction type that may have the main UTXO as input, e.g. redemption.
+	return hasDepositInputs, nil
+}
+
+func (sm *spvMaintainer) isInputCurrentWalletsMainUTXO(
+	fundingTxHash bitcoin.Hash,
+	fundingOutputIndex uint32,
+	walletPublicKeyHash [20]byte,
+) (bool, error) {
+	// Get the transaction the input originated from to calculate the input value.
+	previousTransaction, err := sm.btcChain.GetTransaction(fundingTxHash)
+	if err != nil {
+		return false, fmt.Errorf("failed to get previous transaction: [%v]", err)
+	}
+	fundingOutputValue := previousTransaction.Outputs[fundingOutputIndex].Value
+
+	// Assume the input is the main UTXO and calculate hash.
+	mainUtxoHash := sm.chain.ComputeMainUtxoHash(&bitcoin.UnspentTransactionOutput{
+		Outpoint: &bitcoin.TransactionOutpoint{
+			TransactionHash: fundingTxHash,
+			OutputIndex:     fundingOutputIndex,
+		},
+		Value: fundingOutputValue,
+	})
+
+	// Get the wallet and check if its main UTXO matches the calculated hash.
+	// TODO: `GetWallet` may return an error if the wallet is not found.
+	//       Should we return an error in such case or continue?
```

IMO, we should return an error. Most likely, this is something bad on the infrastructure level (e.g. connection error to Ethereum node). Less likely, this is some terrible bug in the code.

---

Agree. Left the code as is (we return an error if wallet not found) and removed the TODO in ae31a234b0f5ba60571f63479c158d633c15a4eb.